keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,667 @@
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: JavaApplication
+    description: "A Kustomize plugin provider that can handle Java apps"
+    provider: 
+      container: 
+        image: example/module_providers/java:v1.0.0
+```
+
+This will enable a users to use plugin providers with provider information omitted, such as:
+
+```yaml
+# app/kustomization.yaml
+kind: Kustomization
+catalogs:
+  - https://example.com/plugins/catalog.yaml
+resources:
+- javaapp.yaml
+```
+
+```yaml
+# app/javaapp.yaml
+apiVersion: example.com/v1
+kind: JavaApplication
+spec:
+  application: team/my-app
+  version: v1.0
+```
+
+When this Kustomization is processed by `kustomize build`, the referenced catalog (or catalogs) will be used to locate a plugin provider that supports the apiVersion `example.com/v1` and kind `JavaApplication`. If found in one of the referenced catalogs, kustomize can determine the provider configuration without the need for the user to specify it in the kustomization resources directly. The catalogs will be searched in order specified. If more than one catalog defines the target apiVersion and kind, the first will be selected. 
+
+In addition to the new catalog kind, `kustomize build` will accept a repeatable flag `--trusted-plugin-catalog=""`. When present, this flag instructs `kustomize build` to automatically fetch and execute plugins that are defined by the catalog and referenced within the Kustomization, Component or Composition. When a resource is processed by `kustomize build` and a catalog is referenced but not specified using the `--trusted-plugin-catalog=""` flag, an error will occur. Kustomize will provide a built in Catalog for supporting official extensions, published to a well publicized endpoint. This catalog will _NOT_ require the user to explicitly trust it. Users can provide the `apiVersion` and `kind` of the official extensions in kustomize resource and these will be resolved by the official catalog.
+
+In addition to container based plugin providers, the `Catalog` will support discovery of Starlark and Exec based plugin providers, via an HTTP(s), Git, or OCI reference as illustrated below: 
+
+```yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-plugins"
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: GroovyApplication
+    description: "A Kustomize plugin provider that can handle groovy apps"
+    provider:  
+      starlark: https://example.co/module_providers/starlark-func:v1.0.0
+```
+
+This concept can be extended later to support additional plugin provider packaging, but is out of scope for the current proposal.
+
+When HTTP(s) references are used, the HTTP(s) endpoint must support anonymous access for reading resources. Resources will be expected to be stored as a single file, such as `catalog.yaml`. Git or OCI references can be authenticated or anonymous, and will use appropriate configuration from the users file-system. 
+
```

Comment:
I've added stronger wording regarding verification of binary based functions: the catalog must contain a sha256 and they would need to be verified prior to running. I think supporting them remains important, but we could make this a more out of band process (like the kustomize localize command up above) so that users need to run a command that would fetch any providers (Container, exec, whatever) prior to running `kustomize build`?